### PR TITLE
SDN-4024: Add ANP Feature Gate

### PR DESCRIPTION
**Pull request by tssurya**
This PR adds support for feature gates in CNO.
Specifically we want to leverage this for ANP feature
which is in v1alpha1 version. Until it graduates to beta
this feature is TechPreview and hence to be enabled
only on TechPreviewNoUpgrade clusters until we GA this
feature.

```
I0715 23:01:50.451710  903672 operator.go:88] Creating status manager for stand-alone cluster
I0715 23:01:50.451762  903672 operator.go:101] Adding controller-runtime controllers
I0715 23:01:50.451972  903672 operconfig_controller.go:71] SURYA: Feature Gates!!
I0715 23:01:50.453130  903672 operconfig_controller.go:94] SURYA &{0xc000647548  <nil> {0 0} 600000000000 map[] map[] map[]}/4.14.0-0.nightly-2023-07-15-101654/0.0.1-snapshot
I0715 23:01:50.453317  903672 simple_featuregate_reader.go:171] Starting feature-gate-detector
I0715 23:01:50.691526  903672 operconfig_controller.go:118] FeatureGates initialized: knownFeatureGates=[AWSSecurityTokenService AdminNetworkPolicy AdmissionWebhookMatchConditions AlibabaPlatform AutomatedEtcdBackup AzureWorkloadIdentity BuildCSIVolumes CSIDriverSharedResource CloudDualStackNodeIPs DynamicResourceAllocation EventedPLEG ExternalCloudProvider ExternalCloudProviderAzure ExternalCloudProviderExternal ExternalCloudProviderGCP GCPLabelsTags GatewayAPI InsightsConfigAPI MachineAPIOperatorDisableMachineHealthCheckController MachineAPIProviderOpenStack MaxUnavailableStatefulSet NodeSwap OpenShiftPodSecurityAdmission PDBUnhealthyPodEvictionPolicy PrivateHostedZoneAWS RetroactiveDefaultStorageClass RouteExternalCertificate SigstoreImageVerification VSphereStaticIPs]
I0715 23:01:50.691612  903672 event.go:298] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-network-operator", Name:"network-operator", UID:"", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'FeatureGatesInitialized' FeatureGates updated to featuregates.Features{Enabled:[]v1.FeatureGateName{"AlibabaPlatform", "CloudDualStackNodeIPs", "ExternalCloudProviderAzure", "OpenShiftPodSecurityAdmission", "PrivateHostedZoneAWS"}, Disabled:[]v1.FeatureGateName{"AWSSecurityTokenService", "AdminNetworkPolicy", "AdmissionWebhookMatchConditions", "AutomatedEtcdBackup", "AzureWorkloadIdentity", "BuildCSIVolumes", "CSIDriverSharedResource", "DynamicResourceAllocation", "EventedPLEG", "ExternalCloudProvider", "ExternalCloudProviderExternal", "ExternalCloudProviderGCP", "GCPLabelsTags", "GatewayAPI", "InsightsConfigAPI", "MachineAPIOperatorDisableMachineHealthCheckController", "MachineAPIProviderOpenStack", "MaxUnavailableStatefulSet", "NodeSwap", "PDBUnhealthyPodEvictionPolicy", "RetroactiveDefaultStorageClass", "RouteExternalCertificate", "SigstoreImageVerification", "VSphereStaticIPs"}}
I0715 23:01:50.691665  903672 operconfig_controller.go:130] SURYA false
```

Test results:
```
  spec:
    featureSet: TechPreviewNoUpgrade
  status:
    featureGates:
    - disabled:
      - name: EventedPLEG
      - name: MachineAPIOperatorDisableMachineHealthCheckController
      enabled:
      - name: AdminNetworkPolicy --> WOHOHO
      - name: AdmissionWebhookMatchConditions
      - name: AlibabaPlatform
      - name: AutomatedEtcdBackup
      - name: AzureWorkloadIdentity
      - name: BuildCSIVolumes
      - name: CSIDriverSharedResource
      - name: CloudDualStackNodeIPs
      - name: DynamicResourceAllocation
      - name: ExternalCloudProvider
      - name: ExternalCloudProviderAzure
      - name: ExternalCloudProviderExternal
      - name: ExternalCloudProviderGCP
      - name: GCPLabelsTags
      - name: GatewayAPI
      - name: InsightsConfigAPI
      - name: MachineAPIProviderOpenStack
      - name: MaxUnavailableStatefulSet
      - name: NodeSwap
      - name: OpenShiftPodSecurityAdmission
      - name: PDBUnhealthyPodEvictionPolicy
      - name: PrivateHostedZoneAWS
      - name: RetroactiveDefaultStorageClass
      - name: RouteExternalCertificate
      - name: SigstoreImageVerification
      - name: VSphereStaticIPs
```

```
$ oc create -f anp.yaml 
adminnetworkpolicy.policy.networking.k8s.io/pass-example created
[surya@hidden-temple anp]$ oc get anp
NAME           PRIORITY   AGE
pass-example   10         5s
```

```
$ oc logs -n openshift-network-operator network-operator-6db76dbbbc-cgpnj | grep SURYA
I0807 08:48:47.083768       1 operconfig_controller.go:72] SURYA: Feature Gates!!
I0807 08:48:47.084440       1 operconfig_controller.go:95] SURYA &{0xc000d3a168  <nil> {0 0} 600000000000 map[] map[] map[]}/4.14.0-0.ci.test-2023-08-07-083154-ci-ln-qbhvhqb-latest/0.0.1-snapshot
I0807 08:48:47.089247       1 operconfig_controller.go:133] SURYA true
I0807 08:48:48.376462       1 ovn_kubernetes.go:323] SURYA true
I0807 08:49:07.477145       1 ovn_kubernetes.go:323] SURYA true
I0807 08:49:27.455829       1 ovn_kubernetes.go:323] SURYA true
I0807 08:49:47.863910       1 ovn_kubernetes.go:323] SURYA true
I0807 08:50:07.723261       1 ovn_kubernetes.go:323] SURYA true
I0807 08:50:28.086196       1 ovn_kubernetes.go:323] SURYA true
I0807 08:50:47.696208       1 ovn_kubernetes.go:323] SURYA true
I0807 08:51:07.670005       1 ovn_kubernetes.go:323] SURYA true
```
```
$ oc get pods -n openshift-network-operator
NAME                                READY   STATUS    RESTARTS      AGE
network-operator-6db76dbbbc-cgpnj   1/1     Running   1 (92m ago)   96m
```
```
_uuid               : cbc49b09-61cf-4620-a2a2-e09295008e81
acls                : [0315d27b-e3ef-4105-bf63-78a262d8b8b1, cad960fb-f4ed-4967-802f-e3f9c9322235]
external_ids        : {AdminNetworkPolicy=pass-example, name="ANP:pass-example"}
name                : a18043817620427238586
ports               : []
```

```
$ oc logs -n openshift-ovn-kubernetes ovnkube-master-r5p4z ovnkube-master | grep Admin
I0807 08:47:03.219784       1 config.go:2050] Parsed config: {Default:{MTU:1360 RoutableMTU:0 ConntrackZone:64000 EncapType:geneve EncapIP: EncapPort:6081 InactivityProbe:100000 OpenFlowProbe:180 OfctrlWaitBeforeClear:0 MonitorAll:true LFlowCacheEnable:true LFlowCacheLimit:0 LFlowCacheLimitKb:1048576 RawClusterSubnets:10.128.0.0/14/23 ClusterSubnets:[] EnableUDPAggregation:true Zone:global} Logging:{File: CNIFile: Level:4 LogFileMaxSize:100 LogFileMaxBackups:5 LogFileMaxAge:5 ACLLoggingRateLimit:20} Monitoring:{RawNetFlowTargets: RawSFlowTargets: RawIPFIXTargets: NetFlowTargets:[] SFlowTargets:[] IPFIXTargets:[]} IPFIX:{Sampling:400 CacheActiveTimeout:60 CacheMaxFlows:0} CNI:{ConfDir:/etc/cni/net.d Plugin:ovn-k8s-cni-overlay} OVNKubernetesFeature:{EnableAdminNetworkPolicy:true EnableEgressIP:true EgressIPReachabiltyTotalTimeout:1 EnableEgressFirewall:true EnableEgressQoS:true EnableEgressService:true EgressIPNodeHealthCheckPort:9107 EnableMultiNetwork:true E
```
OK NOW let's toggle the knob to see if CNO restarts...(Note that OVNK doesn't have a clean way to disable the feature - user is expected to cleanup all the DB stuff)

err i actually don't know how to toggle post installation :D

### CODE REVIEW

File: pkg/controller/operconfig/operconfig_controller.go [hunk 1]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
+	})
+
+	// By default, this will exit(0) the process if the featuregates ever change to a different set of values.
+	featureGateAccessor := featuregates.NewFeatureGateAccess(
+		desiredVersion, missingVersion,
+		configInformers.Config().V1().ClusterVersions(), configInformers.Config().V1().FeatureGates(),
+		eventRecorder,
+	)
+	go featureGateAccessor.Run(ctx)
+	go configInformers.Start(stopCh)
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:110]: Have you considered using `c.Default().AddCustomInformer`?

File: pkg/controller/operconfig/operconfig_controller.go [hunk 2]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
-		mapper: mgr.GetRESTMapper(),
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager, c cnoclient.Client) (*ReconcileOperConfig, error) {
+	ctx := context.Background()
+	ctx, cancelFn := context.WithCancel(ctx)
+	stopCh := make(chan struct{})
+	go func() {
+		defer cancelFn()
+		<-stopCh
+	}()
+	kubeConfig, err := config.GetConfig()
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:80]: oh thanks pat, I admit my knowledge of all our client stuff in cno is bleaky. changing this to adapt to your suggestion

[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:80]: done!

File: pkg/controller/operconfig/operconfig_controller.go [hunk 3]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get kube config: %v", err)
+	}
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:84]: ha nice

File: pkg/controller/operconfig/operconfig_controller.go (hunk 1 above)
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:110]: no i didn't even know it existed... lemme check

File: bindata/network/ovn-kubernetes/common/003-rbac-controller.yaml [hunk 4]
```diff
@@ -106,6 +106,20 @@ rules:
...
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups: ["policy.networking.k8s.io"]
+  resources:
+  - adminnetworkpolicies/status
+  - baselineadminnetworkpolicies/status
+  verbs:
+  - update
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/003-rbac-controller.yaml:122]: Is cluster-manager updating `adminnetworkpolicies/status`, `baselineadminnetworkpolicies/status` together with ovnkube-node?

File: pkg/controller/operconfig/operconfig_controller.go [hunk 5]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
+		}
+	case <-time.After(1 * time.Minute):
+		return nil, fmt.Errorf("timed out waiting for FeatureGate detection")
+	}
+
+	featureGates, err := featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return nil, err
+	}
+	// example of future featuregate read and usage to set a variable to pass to a controller
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:128]: yes this was just the framework I setup for future things where we populate only things cno cares about into featureGates map. I don't mind using featureGates directly instead whereby we keep track of all things.
changing this to passing featureGates directly..

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml [hunk 6]
```diff
@@ -173,6 +173,20 @@ rules:
   - patch
   - update
   - watch
+- apiGroups: ["policy.networking.k8s.io"]
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:176]: Any specific reason for not putting the RBACs under `OVN_ADMIN_NETWORK_POLICY_ENABLE` if?

File: pkg/network/ovn_kubernetes_test.go [hunk 7]
```diff
@@ -85,9 +86,11 @@ func TestRenderOVNKubernetes(t *testing.T) {
 			},
 		},
 	}
+	featureGatesCNO := make(map[string]bool, 0)
+	featureGatesCNO[string(configv1.FeatureGateAdminNetworkPolicy)] = true
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/network/ovn_kubernetes_test.go:90]: ack, changing this to using featureGates directly based on previous comment

File: pkg/network/render_test.go [hunk 8]
```diff
@@ -386,7 +388,10 @@ func TestRenderUnknownNetwork(t *testing.T) {
 	bootstrapResult, err := Bootstrap(&config, client)
 	g.Expect(err).NotTo(HaveOccurred())
 
-	objs, _, err := Render(prev, bootstrapResult, manifestDir, client)
+	featureGatesCNO := make(map[string]bool, 0)
+	featureGatesCNO[string(configv1.FeatureGateAdminNetworkPolicy)] = true
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/network/render_test.go:392]: hehe good catch, we don't.
removing this.

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml (hunk 6 above)
Comment (tssurya) [reply bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:176]: oh hmm good question. I actually angled to doing this very recently :D forgot to add it to the RBAC stuff. lemme do that

File: bindata/network/ovn-kubernetes/common/003-rbac-controller.yaml (hunk 4 above)
Comment (tssurya) [reply bindata/network/ovn-kubernetes/common/003-rbac-controller.yaml:122]: so this stemmed from confusion - this file should be called `003-rbac-ovnkube-control-plane`? 
s/ovnkube-controller/ovnkube-node/ovnkube-control-plane
changing this to removing perms from this file. control plane doesn't do anything to anp/banp today

File: pkg/controller/operconfig/operconfig_controller.go [hunk 9]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
-		client: c,
-		status: status,
-		mapper: mgr.GetRESTMapper(),
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager, c cnoclient.Client) (*ReconcileOperConfig, error) {
+	ctx := context.Background()
+	ctx, cancelFn := context.WithCancel(ctx)
+	stopCh := make(chan struct{})
+	go func() {
+		defer cancelFn()
+		<-stopCh
```
[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:78]: lol good question dunno.. copied from ingress controller sample code.. lemme check

File: pkg/controller/operconfig/operconfig_controller.go [hunk 10]
```diff
@@ -51,18 +59,82 @@ var ManifestPath = "./bindata"
...
+	go configInformers.Start(stopCh)
+
+	select {
+	case <-featureGateAccessor.InitialFeatureGatesObserved():
+		featureGates, _ := featureGateAccessor.CurrentFeatureGates()
+		if err != nil {
+			return nil, err
+		} else {
+			klog.Infof("FeatureGates initialized: knownFeatureGates=%v", featureGates.KnownFeatures())
+		}
```
Comment (jcaamano) [inline pkg/controller/operconfig/operconfig_controller.go:119]: Do we need this code? We are not setting `err` to anything here

File: pkg/controller/operconfig/operconfig_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply pkg/controller/operconfig/operconfig_controller.go:78]: We could just cancel the context & stop the informer when we have obtained the initial feature gates? It doesn't seem like we are concerned about runtime changes on the enabled feature gates. Should we?

[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:76]: i think we are, it seems feature gates can be enabled and disabled at any time apparently.

File: pkg/controller/operconfig/operconfig_controller.go (hunk 10 above)
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:111]: changed the L114 to actually get the err, good catch!

File: pkg/controller/operconfig/operconfig_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply pkg/controller/operconfig/operconfig_controller.go:76]: So I don't understand. How are we reacting to feature gate config runtime changes on this PR?

[the remark this answers is not in the document]
Comment (kyrtapz) [reply pkg/controller/operconfig/operconfig_controller.go:76]: https://github.com/openshift/cluster-network-operator/pull/1859/commits/01fa8e5a7501143df9952d6d2d45069003263d73#diff-7c8b030b905dedc903155ca81c4f4089318866fece8d863fa881ece4f46f9333R95-R96

the FeatureGate accessor exits the process by default

File: pkg/controller/operconfig/operconfig_controller.go (hunk 1 above)
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:102]: pat and I spoke, this situation will end up in deadlock, the AddCustomInformer doesn't start until we create the newReconciler and this won't be created until the informer has started, so for now we go with goroutine approach.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 11]
```diff
@@ -51,18 +57,76 @@ var ManifestPath = "./bindata"
...
+
+	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events("openshift-network-operator"), "cluster-network-operator", &corev1.ObjectReference{
+		APIVersion: "apps/v1",
+		Kind:       "Deployment",
+		Namespace:  "openshift-network-operator",
+		Name:       "network-operator",
+	})
+
+	// By default, this will exit(0) the process if the featuregates ever change to a different set of values.
+	featureGateAccessor := featuregates.NewFeatureGateAccess(
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:96]: Just a thought but in general this doesn't have to be specific to OperConfig controller, there is a chance that other CNO operators could use the access to featureGates in the future

Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:96]: good point, adding TODO.

File: pkg/controller/operconfig/operconfig_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply pkg/controller/operconfig/operconfig_controller.go:76]: Ah, that was unexpected.

Isn't this code more suited for the `RunOperator` function. We have a context there and we might want other areas of the operator to be aware of feature gates in the future as well.

[the remark this answers is not in the document]
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:76]: we will use wait.NeverStop for now.
and I will add a TODO for also doing this on a more global level.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 12]
```diff
@@ -51,18 +57,76 @@ var ManifestPath = "./bindata"
...
+	case <-time.After(1 * time.Minute):
+		return nil, fmt.Errorf("timed out waiting for FeatureGate detection")
 	}
+
+	featureGates, err := featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return nil, err
+	}
+	// example of future featuregate read and usage to set a variable to pass to a controller
+	featureGatesCNO := make(map[string]bool, 0)
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:121]: `featureGatesCNO` isn't used anywhere now, we should remove that variable together with `adminNetworkPolicyEnabled`?

Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:121]: yes hehe it was a rebase glitch, removed this in my new version.
thanks pat :D

File: pkg/controller/operconfig/operconfig_controller.go [hunk 13]
```diff
@@ -51,18 +57,76 @@ var ManifestPath = "./bindata"
...
+		desiredVersion, missingVersion,
+		configInformers.Config().V1().ClusterVersions(), configInformers.Config().V1().FeatureGates(),
+		eventRecorder,
+	)
+	go featureGateAccessor.Run(ctx)
+	go configInformers.Start(stopCh)
+
+	select {
+	case <-featureGateAccessor.InitialFeatureGatesObserved():
+		featureGates, _ := featureGateAccessor.CurrentFeatureGates()
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:106]: did you mean to do `featureGates, err := featureGateAccessor.CurrentFeatureGates()` instead?

Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:106]: yes already done in my new version: https://github.com/openshift/cluster-network-operator/pull/1859/commits/01fa8e5a7501143df9952d6d2d45069003263d73#r1295715056

File: pkg/controller/operconfig/operconfig_controller.go [hunk 14]
```diff
@@ -51,18 +58,72 @@ var ManifestPath = "./bindata"
...
-	return &ReconcileOperConfig{
-		client: c,
-		status: status,
-		mapper: mgr.GetRESTMapper(),
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager, c cnoclient.Client) (*ReconcileOperConfig, error) {
+	ctx := context.Background()
+	ctx, cancelFn := context.WithCancel(ctx)
+	go func() {
+		defer cancelFn()
+	}()
```
Comment (jcaamano) [inline pkg/controller/operconfig/operconfig_controller.go:76]: I would just have these bits close to  `go featureGateAccessor.Run(ctx)`
And we don't actually need the go routine?

File: pkg/controller/operconfig/operconfig_controller.go [hunk 15]
```diff
@@ -51,18 +58,72 @@ var ManifestPath = "./bindata"
...
+		desiredVersion, missingVersion,
+		configInformers.Config().V1().ClusterVersions(), configInformers.Config().V1().FeatureGates(),
+		eventRecorder,
+	)
+	go featureGateAccessor.Run(ctx)
+	// TODO: 1) If other controllers in CNO also want to use featureGates then we should move this code to outside
+	// operconfig-controller 2) For now we pass the neverStop channel; revisit this logic by trying to use c.Default().AddCustomInformer
+	go configInformers.Start(wait.NeverStop)
+
+	select {
```
Comment (jcaamano) [inline pkg/controller/operconfig/operconfig_controller.go:105]: Perhaps a log that we are waiting for feature gates?

File: pkg/controller/operconfig/operconfig_controller.go (hunk 14 above)
Comment (tssurya) [reply pkg/controller/operconfig/operconfig_controller.go:76]: changing this to `go featureGateAccessor.Run(context.TODO())`

**Comment by jcaamano** [comment]
This one is also permafailing

/override ci/prow/e2e-vsphere-ovn-windows

**Comment by jcaamano** [comment]
I don't think e2e-gcp-sdn failure is related to this PR where the changes are not network type specific. And it has passed before on a previous commit. Checked the logs and CNO does not seem to have any apparent problem related to these changes.

**Comment by jcaamano** [comment]
The job seems to be failing because of an unrelated issue:
'No manifest file name passed' just before starting to run the tests after an apparently successful deployment. The job has already passed once under the current commit. 

/override ci/prow/e2e-metal-ipi-ovn-ipv6

**Pull request merged**